src/Common/src/Internal/Text/Utf8String.cs

```diff
@@ -112,5 +112,41 @@ public bool Equals(Utf8String other)
                 }
             }
         }
+
+        public int CompareTo(object other)
+        {
+            if (!(other is Utf8String))
+                throw new ArgumentException(nameof(other));
+
+            return Compare(this, (Utf8String)other);
+        }
+
+        private static int Compare(Utf8String strA, Utf8String strB)
+        {
+            int length = Math.Min(strA.Length, strB.Length);
+
+            unsafe
+            {
+                fixed (byte* ap = strA._value)
+                fixed (byte* bp = strB._value)
+                {
+                    byte* a = ap;
+                    byte* b = bp;
+
+                    while (length > 0)
+                    {
+                        if (*a != *b)
+                            return *a - *b;
+                        a += 1;
+                        b += 1;
+                        length -= 1;
+                    }
+
+                    // At this point, we have compared all the characters in at least one string.
+                    // The longer string will be larger.
+                    return strA.Length - strB.Length;
```

Comment:
Why use this comparison at the end? Wouldn't it be better to compare the lengths as the first thing, as they are super cheap to compare.

Reply:
Oh, I suppose you're trying to make it so that "Foo" and "Foo1" sort near each other. That's reasonable, although it certainly deserves a comment.

Reply:
We match the semantics of the vanilla `String.Compare` here. I can add a comment.